test/functional/plugin/lsp/snippet_spec.lua

```diff
@@ -1,236 +1,188 @@
 local helpers = require('test.functional.helpers')(after_each)
-local snippet = require('vim.lsp._snippet')
+local snippet = require('vim.lsp._snippet_grammar')
 
 local eq = helpers.eq
 local exec_lua = helpers.exec_lua
 
-describe('vim.lsp._snippet', function()
+describe('vim.lsp._snippet_grammar', function()
   before_each(helpers.clear)
   after_each(helpers.clear)
 
   local parse = function(...)
-    return exec_lua('return require("vim.lsp._snippet").parse(...)', ...)
+    local snippet = exec_lua('return require("vim.lsp._snippet_grammar").parse(...)', ...)
+    return snippet.data.children
   end
 
-  it('should parse only text', function()
+  it('parses only text', function()
     eq({
-      type = snippet.NodeType.SNIPPET,
-      children = {
-        {
-          type = snippet.NodeType.TEXT,
-          raw = 'TE\\$\\}XT',
-          esc = 'TE$}XT',
-        },
-      },
+      { type = snippet.NodeType.Text, data = 'TE$}XT' },
     }, parse('TE\\$\\}XT'))
   end)
 
-  it('should parse tabstop', function()
+  it('parses tabstops', function()
     eq({
-      type = snippet.NodeType.SNIPPET,
-      children = {
-        {
-          type = snippet.NodeType.TABSTOP,
-          tabstop = 1,
-        },
-        {
-          type = snippet.NodeType.TABSTOP,
-          tabstop = 2,
-        },
-      },
+      { type = snippet.NodeType.Tabstop, data = 1 },
+      { type = snippet.NodeType.Tabstop, data = 2 },
     }, parse('$1${2}'))
   end)
 
-  it('should parse placeholders', function()
```

Comment:
I'm not sure of what to do with this test, since it tests something that LSP doesn't support in theory. The snippet grammar says that the value of a placeholder node consists of a single node, but the current parser allows it to be multiple nodes:
```
any         ::= tabstop | placeholder | choice | variable | text
placeholder ::= '${' int ':' any '}'
```

Is that intentional? Should we support this?

Reply:
We only need to support the grammar from the specification. Fine to drop it.

Reply:
on the other hand could there be any problems with supporting it?
Since the goal of this whole snippet thing shouldn't imo only be to support the lsp spec. That's the minimal requirement. If we can make it more extensible "for free" why not do that.

Reply:
> Since the goal of this whole snippet thing shouldn't imo only be to support the lsp spec.

No, that _is_ the goal. The basic user interface may be shared and exposed, but the underlying snippet format will only ever be LSP -- and LSP for us means "the spec, and nothing but the spec".

(And nothing is "free", you always pay a cost in maintenance and documentation.)

Reply:
I think the spec intended that placeholders could have multiple nodes.

For example, the `any-node` to be multiple nodes.
The `local a = $1` will parsed as `any-node` but it will have text-node and tabstop-node.

My decision policy was that the implementation should be supported VSCode behavior as much as possible.

The VSCode allows `local $1 = ${2:test$1test}`.

Reply:
> My decision policy was that the implementation should be supported VSCode behavior as much as possible.

That is different from Neovim's decision policy, though, which is "If VS Code can't stick to their own spec, that's their problem".

(If the spec can be interpreted as allowing multiple nodes, supporting a permissive reading is fine. That's a different matter.)

Reply:
There are many details that are not specified in the LSP specification.
For example, LSP does not determine the cursor position after applying TextEdit.

In such cases, I tried to check the behavior of VSCode as much as possible. (However, I agree that certain "features" of VSCode should not be implemented.)